File: backend-v2/src/zhishi/server/routes/ledger.py
# ruff: noqa: B008
# FastAPI Depends/Query defaults match the existing server route convention.
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from zhishi.domain.ledger import service
from zhishi.domain.ledger.schemas import (
    Currency,
    Direction,
    EntryCreate,
    EntryPage,
    EntryRead,
    EntryReplace,
    LedgerSummary,
    VersionInput,
)
from zhishi.server.deps import get_db

router = APIRouter(prefix="/api/ledger", tags=["ledger"])


def _call(fn, *args, **kwargs):
    try:
        return fn(*args, **kwargs)
    except LookupError as exc:
        raise HTTPException(404, str(exc)) from exc
    except service.LedgerConflict as exc:
        raise HTTPException(409, str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(422, str(exc)) from exc


@router.get("", response_model=EntryPage)
def list_entries(start: date | None = None, end: date | None = None,
                 currency: Currency | None = None, account: str | None = None,
                 direction: Direction | None = None, query: str = "", deleted: bool = False,
                 limit: int = Query(50, ge=1, le=200), offset: int = Query(0, ge=0),
                 db: Session = Depends(get_db)):
    return _call(service.list_entries, db, start=start, end=end, currency=currency,
                 account=account, direction=direction, query=query, deleted=deleted,
                 limit=limit, offset=offset)


@router.get("/summary", response_model=LedgerSummary)
def summary(start: date, end: date, currency: Currency | None = None,
            account: str | None = None, db: Session = Depends(get_db)):
    return _call(service.summary, db, start, end, currency=currency, account=account)


@router.post("", response_model=EntryRead, status_code=201)
def create_entry(payload: EntryCreate, db: Session = Depends(get_db)):
    return service.to_read(_call(service.create_entry, db, payload))


@router.get("/{entry_id}", response_model=EntryRead)
def get_entry(entry_id: int, db: Session = Depends(get_db)):
    return service.to_read(_call(service.get_entry, db, entry_id))


@router.put("/{entry_id}", response_model=EntryRead)
def replace_entry(entry_id: int, payload: EntryReplace, db: Session = Depends(get_db)):
    return service.to_read(_call(service.replace_entry, db, entry_id, payload, payload.version))


@router.delete("/{entry_id}", response_model=EntryRead)
def delete_entry(entry_id: int, version: int = Query(ge=1), db: Session = Depends(get_db)):
    return service.to_read(_call(service.delete_entry, db, entry_id, version))


@router.post("/{entry_id}/restore", response_model=EntryRead)
def restore_entry(entry_id: int, payload: VersionInput, db: Session = Depends(get_db)):
    return service.to_read(_call(service.restore_entry, db, entry_id, payload.version))

File: frontend-v2/app/src/api/ledger.ts
import type { components } from './contracts/rest'
import { http, request } from './http'

type Schemas = components['schemas']
export type LedgerEntry = Schemas['EntryRead']
export type LedgerInput = Schemas['EntryCreate']
export type LedgerPage = Schemas['EntryPage']
export type LedgerSummary = Schemas['LedgerSummary']
export type LedgerCurrency = LedgerEntry['currency']

export const currencies: LedgerCurrency[] = ['CNY', 'USD', 'EUR', 'GBP', 'HKD', 'JPY']
export function monthRange(month: string): { start: string; end: string } {
  if (!/^\d{4}-(0[1-9]|1[0-2])$/.test(month)) throw new Error('请选择有效月份')
  const [year, m] = month.split('-').map(Number) as [number, number]
  const last = new Date(year, m, 0).getDate()
  return { start: `${month}-01`, end: `${month}-${last}` }
}
export function listLedger(query: { start: string; end: string; currency?: LedgerCurrency;
  account?: string; query?: string; deleted?: boolean; offset?: number }): Promise<LedgerPage> {
  return http.get('/api/ledger', { ...query, limit: 50 })
}
export function ledgerSummary(query: { start: string; end: string; currency?: LedgerCurrency;
  account?: string }): Promise<LedgerSummary> {
  return http.get('/api/ledger/summary', query)
}
export const readEntry = (id: number): Promise<LedgerEntry> => http.get(`/api/ledger/${id}`)
export const createEntry = (input: LedgerInput): Promise<LedgerEntry> => http.post('/api/ledger', input)
export const replaceEntry = (id: number, input: Schemas['EntryReplace']): Promise<LedgerEntry> => http.put(`/api/ledger/${id}`, input)
export const deleteEntry = (entry: LedgerEntry): Promise<LedgerEntry> => request(`/api/ledger/${entry.id}`, { method: 'DELETE', query: { version: entry.version } })
export const restoreEntry = (entry: LedgerEntry): Promise<LedgerEntry> => http.post(`/api/ledger/${entry.id}/restore`, { version: entry.version })
